Update image-button-non-empty-accessible-name-59796f.md

giacomo-petri opened the pull request:
Added:
* failed example 3
* updated "Accessibility Support" section

Closes issue: #1861

Need for Call for Review:
This will require a 1 week Call for Review 

Note: Sorry if I made mistakes while creating the pull request, this is my first pull request and first time I contribute to a GitHub project.

---

## Pull Request Etiquette

### **When creating PR:**

- [ ] Make sure you're requesting to **pull a branch** (right side) to the `develop` branch (left side).
- [ ] Make sure you do not remove the "How to Review and Approve" section in your pull request description

### **After creating PR:**

- [ ] Add yourself (and co-authors) as "Assignees" for PR.
- [ ] Add label to indicate if it's a `Rule`, `Definition` or `Chore`.
- [ ] [Link the PR to any issue it solves](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue). This will be done automatically by [referencing the issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) at the top of this comment in the indicated place.
- [ ] Optionally request feedback from anyone in particular by assigning them as "Reviewers".

### **When merging a PR:**

- [ ] Close any issue that the PR resolves. This will happen automatically upon merging if the PR was correctly linked to the issue, e.g. by referencing the issue at the top of this comment.

## How to Review And Approve

- Go to the “Files changed” tab
- Here you will have the option to leave comments on different lines.
- Once the review is completed, find the “Review changes” button in the top right, select “Approve” (if you are really confident in the rule) or "Request changes" and click “Submit review”.
- Make sure to also review the proposed Call for Review period. In case of disagreement, the longer period wins.

giacomo-petri committed:
Update image-button-non-empty-accessible-name-59796f.md

Added:
* failed example 3
* updated "Accessibility Support" section

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/act-rules/act-rules.github.io?pullRequest=1865) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/act-rules/act-rules.github.io?pullRequest=1865) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/act-rules/act-rules.github.io?pullRequest=1865) it.</sub>

carlosapaduarte commented on _rules/image-button-non-empty-accessible-name-59796f.md:
```diff
@@ -56,6 +56,8 @@ Each target element has an [accessible name][] that is neither empty (`""`), nor
 
 There is a known combination of a popular browser and assistive technology that does not by default support `title` as an [accessible name][].
 
+The image button with non-empty `title` attribute but empty `alt` attribute results in a button with empty accessible name on many browsers. On Firefox, instead, the `title` attribute is used as fallback, resulting in a button with `title` attribute value as accessible name. 
```
I think we should try to write this without naming any specific browsers (like we do in the previous sentence).

giacomo-petri replied:
Thanks Carlos,
updated using a generic _known popular browser_  as above.

carlosapaduarte reviewed (changes requested):
No mistakes when creating the PR! 👍 

I just left a request to do the updated accessibility support section without naming any specific browsers.

giacomo-petri committed:
Update image-button-non-empty-accessible-name-59796f.md

carlosapaduarte commented on _rules/image-button-non-empty-accessible-name-59796f.md:
```diff
@@ -56,6 +56,8 @@ Each target element has an [accessible name][] that is neither empty (`""`), nor
 
 There is a known combination of a popular browser and assistive technology that does not by default support `title` as an [accessible name][].
 
+The image button with non-empty `title` attribute but empty `alt` attribute results in a button with empty accessible name on many browsers. There is a known popular browser instead, that uses the `title` attribute as fallback even with empty `alt`, resulting in a button with `title` attribute value as accessible name. 
```
```suggestion
The image button with non-empty `title` attribute but empty `alt` attribute results in a button with empty accessible name on many browsers. There is a known popular browser that uses the `title` attribute as fallback even with empty `alt`, resulting in a button with `title` attribute value as accessible name. 
```

carlosapaduarte reviewed (changes requested):
Editorial suggestion

giacomo-petri committed:
Update _rules/image-button-non-empty-accessible-name-59796f.md

Co-authored-by: Carlos Duarte <carlosapaduarte@gmail.com>

giacomo-petri committed:
Update image-button-non-empty-accessible-name-59796f.md

My mistake; the resulting accessible name of Failed example 3 is empty (except for Firefox), not default "submit query".

Jym77 commented on _rules/image-button-non-empty-accessible-name-59796f.md:
```diff
@@ -56,6 +56,8 @@ Each target element has an [accessible name][] that is neither empty (`""`), nor
 
 There is a known combination of a popular browser and assistive technology that does not by default support `title` as an [accessible name][].
 
+The image button with non-empty `title` attribute but empty `alt` attribute results in a button with empty accessible name on many browsers. There is a known popular browser that uses the `title` attribute as fallback even with empty `alt`, resulting in a button with `title` attribute value as accessible name. 
```
```suggestion
An image button with non-empty `title` attribute but empty `alt` attribute results in a button with an empty accessible name on many browsers. There is a known popular browser that uses the `title` attribute as fallback even with empty `alt`, resulting in a button with `title` attribute value as accessible name. 
```

Jym77 commented on _rules/image-button-non-empty-accessible-name-59796f.md:
```diff
@@ -128,6 +130,14 @@ The image button has an empty `alt` attribute, and no other attributes that can
 
 #### Failed Example 3
 
+The image button has a non-empty `title` attribute, but an empty `alt` attribute. This gives the button an empty [accessible name][].
```
```suggestion
This image button has a non-empty `title` attribute, but an empty `alt` attribute. This gives the button an empty [accessible name][].
```

Jym77 reviewed (dismissed):
Typos.

giacomo-petri committed:
Update _rules/image-button-non-empty-accessible-name-59796f.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

giacomo-petri committed:
Update _rules/image-button-non-empty-accessible-name-59796f.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

Jym77 committed:
Merge branch 'develop' into giacomo-petri-image-button-non-empty-accessible-name

scottaohara commented on _rules/image-button-non-empty-accessible-name-59796f.md:
```diff
@@ -56,6 +56,8 @@ Each target element has an [accessible name][] that is neither empty (`""`), nor
 
 There is a known combination of a popular browser and assistive technology that does not by default support `title` as an [accessible name][].
 
+An image button with non-empty `title` attribute but empty `alt` attribute results in a button with an empty accessible name on many browsers. There is a known popular browser that uses the `title` attribute as fallback even with empty `alt`, resulting in a button with `title` attribute value as accessible name. 
```
an image and an image button are different things.  they have two different naming algorithms for a reason.  why are they being merged into one test?

giacomo-petri replied:
I agree with you; we were almost talking about the need of a clearer definition for "usable text string" in accessible name computation rules.

That said, this rule is specific for image button elements (we didn't merge image and image button elements int one single test).

As per [Inapplicable Example 4](https://www.w3.org/WAI/content-assets/wcag-act-rules/testcases/59796f/ebd0080bacb8debc7ad069072240657df38c3e2c.html)
> The img element is not a user interface component, and so is not tested for [Success Criterion 4.1.2 Name, Role, Value](https://www.w3.org/TR/WCAG21/#name-role-value).

that confirms that images are not part of this rule.

Jym77 replied:
In light of @scottaohara clarification at https://github.com/w3c/html-aam/issues/414#issuecomment-1178384158, I think we can rephrase this in the line of
> The naming algorithm (+link to relevant HTML AAM part) uses the first non-empty name, but some AU/AT stop at the first existing one, even if empty.

This would make it clear that it is not just some browsers quirks deciding what to do, but an actual bug in some implementation (and would explain why our examples are built following specs and may behave differently on some browsers).

---
Maybe we can also add a background note in the line of:
> Contrarily to images, empty `alt` does not make image button decorative. Image buttons are button and therefore exposed as interactive elements.

which is one on these things that is obvious once stated but maybe not so obvious before seeing it once…

scottaohara commented on _rules/image-button-non-empty-accessible-name-59796f.md:
````diff
@@ -128,6 +130,14 @@ The image button has an empty `alt` attribute, and no other attributes that can
+This image button has a non-empty `title` attribute, but an empty `alt` attribute. This gives the button an empty [accessible name][].
+
+```html
+<input type="image" src="/test-assets/shared/search-icon.svg" title="Search" alt="" />
````
firefox says search button
chrome has no name (bug)
safari says submit button

firefox is correct here
chrome is wrong by providing no name
safari skips title but at least provides the fallback text.

giacomo-petri replied:
In my opinion this might be subjective (and that's probably also why browsers render this example in different ways).

From [4.3.1 input type="image" Accessible Name Computation](https://www.w3.org/TR/html-aam/#input-type-image-accessible-name-computation)

> 4.3.1 input type="image" Accessible Name Computation
    1. If the control has an [aria-label](https://www.w3.org/TR/wai-aria-1.2/#aria-label) or an [aria-labelledby](https://www.w3.org/TR/wai-aria-1.2/#aria-labelledby) attribute the [accessible name](https://www.w3.org/TR/accname-1.2/#dfn-accessible-name) is to be calculated using the algorithm defined in [Accessible Name and Description: Computation and API Mappings](https://www.w3.org/TR/accname-1.2/).
    2. Otherwise use alt attribute.
    3. Otherwise use title attribute.
    4. Otherwise if the previous steps do not yield a usable text string, the [accessible name](https://www.w3.org/TR/accname-1.2/#dfn-accessible-name) is a localized string of the word "Submit Query".
If none of the above yield a usable text string there is no [accessible name](https://www.w3.org/TR/accname-1.2/#dfn-accessible-name).

In our example the alt attribute has been clearly defined as empty (alt=""). Depending on interpretations, we might stop at point 2, considering the alt="" as a choice to make the image decorative, or we can move forward reviewing the title.

Even more important is that for images, the HTML Accessibility API Mapping contemplate (rightly) the specific scenario ([4.10 img Element](https://www.w3.org/TR/html-aam/#img-element-accessible-name-computation))

> 4.10.1 img Element Accessible Name Computation
    1. If the img element has an [aria-label](https://www.w3.org/TR/wai-aria-1.2/#aria-label) or an [aria-labelledby](https://www.w3.org/TR/wai-aria-1.2/#aria-labelledby) attribute the [accessible name](https://www.w3.org/TR/accname-1.2/#dfn-accessible-name) is to be calculated using the algorithm defined in [Accessible Name and Description: Computation and API Mappings](https://www.w3.org/TR/accname-1.2/).
    2. Otherwise use alt attribute, even if its value is the empty string.
    NOTE
    An img with an alt attribute whose value is the empty string is mapped to the [presentation](https://www.w3.org/TR/core-aam-1.2/#role-map-presentation) role. It has no accessible name.
    3. Otherwise, if there is no alt attribute use the title attribute.
    4. Otherwise there is no [accessible name](https://www.w3.org/TR/accname-1.2/#dfn-accessible-name).

Combining these two items, might result unclear which are expectations.

Interesting also the example provided by @Jym77 when images are not loaded.

giacomo-petri replied:
In addition, if these are the expectations, it is necessary that all browsers move in this direction; right now, this example might be an important barrier for some user depending on the technology the user is using.

Jym77 replied:
In the light of the recent clarifications, this should be a Passed Example indeed (since the empty `alt` does not provide a "usable string").
Description can say that the image button has name "Search", but some UA/AT (incorrectly) either use the empty `alt` or the fallback "Submit query".

giacomo-petri committed:
Update image-button-non-empty-accessible-name-59796f.md

giacomo-petri committed:
Update image-button-non-empty-accessible-name-59796f.md

Jym77 committed:
Merge branch 'develop' into giacomo-petri-image-button-non-empty-accessible-name

Jym77 commented on _rules/image-button-non-empty-accessible-name-59796f.md:
```diff
@@ -56,8 +56,13 @@ Each target element has an [accessible name][] that is neither empty (`""`), nor
+
 ## Background
 
+Empty `alt` does not provide a "usable string" for `input` element with a `type` [attribute value][] of `image`.
+Therefore, contrarily to images, empty `alt` does not make image button decorative. Image buttons have a button role and therefore are exposed as interactive elements.
```
```suggestion
Contrarily to images, an empty `alt` attribute does not make image button decorative; image buttons have a button role and are therefore exposed as interactive elements. Consequently, an empty `alt` attribute does not provide a "usable string" for image buttons and the computation default to other means of providing a name.
```

Jym77 commented on _rules/image-button-non-empty-accessible-name-59796f.md:
```diff
@@ -56,8 +56,13 @@ Each target element has an [accessible name][] that is neither empty (`""`), nor
 
 There is a known combination of a popular browser and assistive technology that does not by default support `title` as an [accessible name][].
 
+The [input type="image" Accessible Name Computation algorithm](https://www.w3.org/TR/html-aam/#input-type-image-accessible-name-computation) uses the first non-empty name, but some AU/AT stop at the first existing one, even if empty.
```
```suggestion
The [input type="image" Accessible Name Computation algorithm](https://www.w3.org/TR/html-aam/#input-type-image-accessible-name-computation) uses the first non-empty name, but some user agents and assistive technologies combinations stop at the first existing one, even if empty.
```

Jym77 reviewed (changes requested):
Polishing a bit the notes.

Jym77 committed:
Merge branch 'develop' into giacomo-petri-image-button-non-empty-accessible-name

giacomo-petri committed:
Apply suggestions from code review

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

giacomo-petri committed:
Create error-field-is-not-ifentified-using-color-differences-only

giacomo-petri committed:
Delete error-field-is-not-ifentified-using-color-differences-only

Jym77 commented on _rules/image-button-non-empty-accessible-name-59796f.md:
```diff
@@ -56,12 +56,11 @@ Each target element has an [accessible name][] that is neither empty (`""`), nor
 
-Empty `alt` does not provide a "usable string" for `input` element with a `type` [attribute value][] of `image`.
-Therefore, contrarily to images, empty `alt` does not make image button decorative. Image buttons have a button role and therefore are exposed as interactive elements.
+Contrarily to images, an empty `alt` attribute does not make image button decorative; image buttons have a button role and are therefore exposed as interactive elements. Consequently, an empty `alt` attribute does not provide a "usable string" for image buttons and the computation default to other means of providing a name.
```
```suggestion
Contrarily to images, an empty `alt` attribute does not make image button decorative; image buttons have a button role and are therefore exposed as interactive elements. Consequently, an empty `alt` attribute does not provide a "usable string" for image buttons and the computation defaults to other means of providing a name.
```

Jym77 reviewed (changes requested):
Fixing a typo i my previous suggestion 🙈

giacomo-petri committed:
Update _rules/image-button-non-empty-accessible-name-59796f.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

Jym77 committed:
Merge branch 'develop' into giacomo-petri-image-button-non-empty-accessible-name

WilcoFiers commented on _rules/image-button-non-empty-accessible-name-59796f.md:
```diff
@@ -56,8 +56,12 @@ Each target element has an [accessible name][] that is neither empty (`""`), nor
+
 ## Background
 
+Contrarily to images, an empty `alt` attribute does not make image button decorative; image buttons have a button role and are therefore exposed as interactive elements. Consequently, an empty `alt` attribute does not provide a "usable string" for image buttons and the computation defaults to other means of providing a name.
```
1. I don't think you should use `Contrarily to images`, you're talking specifically about the `img` element. There are different types of images in HTML.
2. I'm not sure what that second sentence means. Can you make it clearer?

giacomo-petri replied:
1. I agree with the 1st item. I'll update "images" with `img` elements.

----

2. About the second sentence
> Consequently, an empty `alt` attribute does not provide a "usable string" for image buttons and the computation defaults to other means of providing a name.

it refers to the previous adjacent Accessibility support paragraph

> The [input type="image" Accessible Name Computation algorithm](https://www.w3.org/TR/html-aam/#input-type-image-accessible-name-computation) uses the first non-empty name, but some user agents and assistive technologies combinations stop at the first existing one, even if empty.

Per [input type="image" Accessible Name Computation algorithm](https://www.w3.org/TR/html-aam/#input-type-image-accessible-name-computation), if the alt attribute value consists in a non-usable string the title attribute should be considered, and again, if the title attribute value consists in a non-usable string, the accessible name is a localized string of the word "Submit Query".

For the sake of clarity I'll update 
> Consequently, an empty `alt` attribute does not provide a "usable string" for image buttons and the computation defaults to other means of providing a name.

in 

> Consequently, an empty `alt` attribute does not provide a "usable string" for image buttons and the computation defaults to other means of providing a name, as defined in [input type="image" Accessible Name Computation algorithm](https://www.w3.org/TR/html-aam/#input-type-image-accessible-name-computation).

Does it make sense?

WilcoFiers commented on _rules/image-button-non-empty-accessible-name-59796f.md:
````diff
@@ -108,6 +112,14 @@ The image button has an [accessible name][] through the `aria-labelledby` attrib
 <div id="id1">Search</div>
 ```
 
+#### Passed Example 5
+
+This image button has an empty `alt` attribute, but a non-empty `title` attribute. This gives the button an non-empty [accessible name][] through the `title` attribute.
+
+```html
+<input type="image" src="/test-assets/shared/search-icon.svg" title="Search" alt="" />
+```
+
````
Since this does indeed cause significant accessibility support issues, I don't think we can include it as a passing example. It's perfectly understandable / correct for testers to fail such image buttons.

```suggestion
```

giacomo-petri replied:
My idea was to clearly define a way to test this scenario, showing a pass/fail example to better understand how to tackle it. It was a real case found in a website but the definition was ambiguous, without clearly explaining if it was a pass or a fail.

If specs clearly define it as a pass, in my opinion we should reiterate our feedback, asking techs to support it as per specs, adding maybe a clear note in Accessibility Support section.

giacomo-petri committed:
Update image-button-non-empty-accessible-name-59796f.md

giacomo-petri committed:
Update image-button-non-empty-accessible-name-59796f.md

giacomo-petri committed:
Update _rules/image-button-non-empty-accessible-name-59796f.md

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>

giacomo-petri commented:
As per our conversation, I've removed Passed example 5 and I've opened an issue about how to handle those cases: #1954

Jym77 committed:
Merge branch 'develop' into giacomo-petri-image-button-non-empty-accessible-name

WilcoFiers committed:
Merge branch 'develop' into giacomo-petri-image-button-non-empty-accessible-name
